components/mailchimp/mailchimp.app.js

```diff
@@ -0,0 +1,369 @@
+        intervalSeconds: 60
+      },
+    },
+  },
+  methods: {
+    _authToken() {
+      return this.$auth.oauth_access_token;
+    },
+    _server() {
+      return this.$auth.dc;
+    },
+    _initMailchimpClient() {
+      mailchimp.setConfig({
+        accessToken: this._authToken(),
+        server: this._server(),
+      });
+      return mailchimp;
+    },
+    async _withRetries(apiCall) {
+      const retryOpts = {
+        retries: 5,
+        factor: 2,
+      };
+      return retry(async (bail) => {
+        try {
+          return await apiCall();
+        } catch (err) {
+          const statusCode = get(err, [
+            "response",
+            "status",
+          ]);
+          if (!this._isRetriableStatusCode(statusCode)) {
+            bail(`
+              Unexpected error (status code: ${statusCode}):
+              ${JSON.stringify(err.response)}
+            `);
+          }
+          console.warn(`Temporary error: ${err.message}`);
+          throw err;
+        }
+      }, retryOpts);
+    },
+    _statusIsSent(status) {
+      return [
+        "status",
+      ].includes(status);
+    },
+    /**
+     * Create a webhook on the specified Audience List.
+     * @params {String} listId - The unique ID that identifies the Audience List you would like to
+     * create a webhook on.
+     * @params {String} webhookUrl - URL listening for the webhook event.
+     * @returns {_links: array, events: object, id: string, list_id: string, sources: object, url:
+     * string} Object with array of links for the webhook manipulation through the Mailchimp API,
+     * an object with flags of the Mailchimp Marketing indicates the events being watched for, an
+     * unique ID of the webhoom, the unique ID which identifies the Audience List where the webhoook
+     * was created, a `sources` object indicating whether the sources for the event are an admin
+     * user, an audience subscriber, or via the API, the URL registered for the webhook event.
+     */
+    async createWebhook(listId, webhookUrl, events, sources) {
+      const mailchimp = this._initMailchimpClient();
+      return await this._withRetries(() =>
+        mailchimp.lists.createListWebhook(listId, {
+          url: webhookUrl,
+          events,
+          sources,
+        }));
+    },
+    /**
+     * Deletes a webhook on the specified Audience List.
+     * @params {String} server - The server prefix used to access the connected Maichimp account.
+     * @params {String} listId - The unique ID that identifies the Audience List containing the
+     * webhook to delete.
+     * @params {String} webhookId - The unique ID of the webhook you'd like to delete.
+     * @returns {} - This method returns an empty object.
+     */
+    async deleteWebhook(server, listId, webhookId) {
+      const mailchimp = this._initMailchimpClient(server);
+      return await this._withRetries(() =>
+        mailchimp.lists.deleteListWebhook(listId, webhookId));
```

Comment:
We'll need to also remove the `server` parameter from the `deleteWebhook` function for the [`deactivate` hook](https://github.com/PipedreamHQ/pipedream/blob/86ef7047bc822d0fd4a8da0c8daf91d256252d01/components/mailchimp/sources/common/http-based.js#L46) to run successfully and allow `pd dev` to update the source.

```suggestion
    async deleteWebhook(listId, webhookId) {
      const mailchimp = this._initMailchimpClient();
      return await this._withRetries(() =>
        mailchimp.lists.deleteListWebhook(listId, webhookId));
```

Reply:
Yes, I had this on my backlog.
btw, @dylburger , something I noticed regarding an account's Mailchimp server, is that this is available via the `$.auths` variable when you are developing components. however, when connecting Mailchimp to run node.js (i.e legacy actions), the `auths` variable doesn't include it.

Reply:
This previously has been addressed.
A `_server()` method been defined that returns the server from the `$.auths` object.